pkg/cloudprovider/providers/azure/azure_loadbalancer.go

```diff
@@ -803,24 +698,107 @@ func (az *Cloud) reconcileLoadBalancer(lb network.LoadBalancer, fipConfiguration
+				}
+			}
+			if err != nil {
+				return nil, err
+			}
+
+		} else {
+			glog.V(3).Infof("ensure(%s): lb(%s) - updating", serviceName, lbName)
+			az.operationPollRateLimiter.Accept()
+			glog.V(10).Infof("LoadBalancerClient.CreateOrUpdate(%q): start", lbName)
+			respChan, errChan := az.LoadBalancerClient.CreateOrUpdate(az.ResourceGroup, lbName, *lb, nil)
+			resp := <-respChan
+			err := <-errChan
+			glog.V(10).Infof("LoadBalancerClient.CreateOrUpdate(%q): end", lbName)
+			if az.CloudProviderBackoff && shouldRetryAPIRequest(resp.Response, err) {
+				glog.V(2).Infof("ensure(%s) backing off: lb(%s) - updating", serviceName, lbName)
+				retryErr := az.CreateOrUpdateLBWithRetry(*lb)
+				if retryErr != nil {
+					glog.V(2).Infof("ensure(%s) abort backoff: lb(%s) - updating", serviceName, lbName)
+					return nil, retryErr
+				}
+			}
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if wantLb && nodes != nil {
+		// Add the machines to the backend pool if they're not already
+		availabilitySetName := az.mapLoadBalancerNameToAvailabilitySet(lbName, clusterName)
+		hostUpdates := make([]func() error, len(nodes))
+		for i, node := range nodes {
+			localNodeName := node.Name
+			f := func() error {
+				err := az.ensureHostInPool(serviceName, types.NodeName(localNodeName), lbBackendPoolID, availabilitySetName)
+				if err != nil {
+					return fmt.Errorf("ensure(%s): lb(%s) - failed to ensure host in pool: %q", serviceName, lbName, err)
+				}
+				return nil
+			}
+			hostUpdates[i] = f
+		}
+
+		errs := utilerrors.AggregateGoroutines(hostUpdates...)
+		if errs != nil {
+			return nil, utilerrors.Flatten(errs)
+		}
+	}
+
+	glog.V(2).Infof("ensure(%s): lb(%s) finished", serviceName, lbName)
+	return lb, nil
 }
 
 // This reconciles the Network Security Group similar to how the LB is reconciled.
 // This entails adding required, missing SecurityRules and removing stale rules.
-func (az *Cloud) reconcileSecurityGroup(sg network.SecurityGroup, clusterName string, service *v1.Service, lbIP *string, wantLb bool) (network.SecurityGroup, bool, error) {
+func (az *Cloud) reconcileSecurityGroup(clusterName string, service *v1.Service, lbStatus *v1.LoadBalancerStatus, wantLb bool) (*network.SecurityGroup, error) {
 	serviceName := getServiceName(service)
+	glog.V(5).Infof("ensure(%s): START clusterName=%q lbName=%q", serviceName, clusterName)
+
 	var ports []v1.ServicePort
 	if wantLb {
 		ports = service.Spec.Ports
 	} else {
 		ports = []v1.ServicePort{}
 	}
+	az.operationPollRateLimiter.Accept()
+	glog.V(10).Infof("SecurityGroupsClient.Get(%q): start", az.SecurityGroupName)
+	sg, err := az.SecurityGroupsClient.Get(az.ResourceGroup, az.SecurityGroupName, "")
+	glog.V(10).Infof("SecurityGroupsClient.Get(%q): end", az.SecurityGroupName)
+	if err != nil {
+		return nil, err
+	}
 
 	destinationIPAddress := ""
 	if wantLb {
+		// Get lbIP since we make up NSG rules based on ingress IP
+		lbIP := &lbStatus.Ingress[0].IP
 		if lbIP == nil {
```

Comment:
Not sure how this can be nil when it comes from an `&` expression on the previous line, unless this is a special Go behaviour (e.g. abuse of syntax to propagate nil if `lbStatus` is nil or something?)?

Reply:
Synced offline. Ivan will help to address this, along with his merging.